### PR TITLE
Log Likelihood of a Tree

### LINKED ISSUE #19: Add the log-probability function
Add the log-probability function, which can be used for plots and for the MCMC inference.

Sprint: #18 
Addressed in a PR: #17

### PR DESCRIPTION
**Pull request by gordonkoehn**
Implementing the `logprobability_fn`: function by taking a tree and returning its log-probability
          (up to the additive (log-)normalization constant).
          
As defined by Eqn. 13 in the SCITE paper.

Taking care of numerical stability using the  `logsumexp` trick.

Consider using `jnp.einsum`.

Resolves #19

### CODE REVIEW

File: src/pyggdrasil/tree_inference/_tree.py [hunk 1]
```diff
@@ -167,7 +166,7 @@ def body(carry, _):
     return adj_matrix_exp
 
 
-def _get_ancestor_matrix(adj_matrix: Array, n: int):
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_tree.py:170]: realized 'n' is inferred from adj_mat anyway

File: src/pyggdrasil/tree_inference/_logprob.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    mutation_likelihood = jnp.where(mask, 1 - alpha, mutation_likelihood)
+    # P(D_{ij} = 0| E_{ik} = 1) = beta
+    mask = (mutation_tensor == 0) & (ancestor_tensor == 1)
+    mutation_likelihood = jnp.where(mask, beta, mutation_likelihood)
+    # P(D_{ij} = 1| E_{ik} = 0) = alpha
+    mask = (mutation_tensor == 1) & (ancestor_tensor == 0)
+    mutation_likelihood = jnp.where(mask, alpha, mutation_likelihood)
+    # P(D_{ij} = 1| E_{ik} = 1) = 1 - beta
+    mask = (mutation_tensor == 1) & (ancestor_tensor == 1)
+    mutation_likelihood = jnp.where(mask, 1 - beta, mutation_likelihood)
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_logprob.py:141]: Does this look right? This is how I understand the math of $P(D_{ij} | A(T)_{i \sigma_j})$. It works as expected, but my intuition of what the operation represents has left me in the 3rd dimension.  These are my workings: 
[SCITE_mutation_likelihood.pdf](https://github.com/cbg-ethz/PYggdrasil/files/11289383/SCITE_mutation_likelihood.pdf)

File: src/pyggdrasil/tree_inference/_tree.py (hunk 1 above)
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_tree.py:170]: Removed unnecessary argument.

File: tests/tree_inference/test_logprob.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+# _logprob.py
+
+import jax.numpy as jnp
+
+import pyggdrasil.tree_inference._logprob as logprob
+from pyggdrasil.tree_inference._tree import Tree
+import pyggdrasil.tree_inference._tree as tr
+
+
+def test_compute_mutation_likelihood():
```
Comment (gordonkoehn) [inline tests/tree_inference/test_logprob.py:11]: Applied example as in 
[SCITE_mutation_likelihood.pdf](https://github.com/cbg-ethz/PYggdrasil/files/11289403/SCITE_mutation_likelihood.pdf)

File: tests/tree_inference/test_logprob.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
+"""Tests of the log-probability calculations."""
```
Comment (gordonkoehn) [inline tests/tree_inference/test_logprob.py:1]: Any more ideas how to test `log_prob` ?

**Review by gordonkoehn — COMMENTED** [review]
Implemented sums and logsumexp as discussed. My concern is with how I calculate $P(D_{ij}|A(T)_{i ~\sigma_j})$.  The respective function,`mutation_likelihood`, does what I expect and passes all my tests, but I am not entirely convinced this is what is meant. I lost intuition of what the operation means.

File: src/pyggdrasil/tree_inference/_logprob.py (hunk 2 above)
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_logprob.py:141]: It looks very good to me! :slightly_smiling_face: 
(But I'm not 100% sure, so it's good that you confirmed on simple examples that it works as expected).

I think you have a tensor which dimensions $ijk$ represent $P(D_ij | A(T)_{ik}$. With this you're good to go with summation (`logsumexp` and `einsum`) as you wrote above.

By the way, a nice trick to do here: if you have the mutation profile of a *single* cell (so that $m=1$ and $j=0$ is the only value index $j$ can take), you have a two-dimensional array which is easier to understand. Then, if you "clone" the cell into $n$ cells the loglikelihood should multiply by a factor of $n$.

Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_logprob.py:141]: Oh yes - of course. Just worked through in 2D. Got the intuition back :) Will write a dynamic 2D test and submit for your review if it passes this.

File: tests/tree_inference/test_logprob.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            [alpha, alpha, 1 - beta, alpha],
+        ]
+    )
+
+    p_mat_true = get_mutation_likelihood(tree, data_true, (alpha, beta))
+    p_mat_false = get_mutation_likelihood(tree, data_false, (alpha, beta))
+
+    assert jnp.all(p_mat_true == p_true)
+    assert jnp.all(p_mat_false == p_false)
+    assert jnp.einsum("ik->", p_mat_true) > jnp.einsum("ik->", p_mat_false)
```
Comment (gordonkoehn) [inline tests/tree_inference/test_logprob.py:124]: Added this 2D test function as suggested. It used only one cell m = 1. Works just fine. For my own record, this is the example 
[2D_Example_MuationLikelihood_2023-04-24.pdf](https://github.com/cbg-ethz/PYggdrasil/files/11308293/2D_Example_MuationLikelihood_2023-04-24.pdf)

**Review by gordonkoehn — COMMENTED** [review]
Having added the 2D test, I am now confident the 'log_probability' is calculated correctly and I have regained my intuition about it. Ready to merge.

File: src/pyggdrasil/tree_inference/_logprob.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    log_mutation_likelihood = jnp.log(mutation_likelihood)
+
+    return log_mutation_likelihood
+
+
+def _mutation_likelihood(
+    mutation_matrix: MutationMatrix,
+    ancestor_matrix: AncestorMatrix,
+    theta: tuple[float, float],
+) -> Mutation_Likelihood:
```
Comment (pawel-czyz) [inline src/pyggdrasil/tree_inference/_logprob.py:93]: (Not anything to be acted on as it's an unimportant detail – feel free to resolve.)

The `snake_case` is in Python usually reserved for functions. Classes typically follow `CamelCase`. (Which is just a convention, but perhaps may give you +1 at the coding interviews :sweat_smile:)

File: src/pyggdrasil/tree_inference/_logprob.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    ancestor_mat = tr._get_ancestor_matrix(tree.tree_topology)
+    # get mutation likelihood
+    mutation_likelihood = _mutation_likelihood(mutation_mat, ancestor_mat, theta)
+    # get log of it
+    log_mutation_likelihood = jnp.log(mutation_likelihood)
+
+    return log_mutation_likelihood
+
+
+def _mutation_likelihood(
```
Comment (pawel-czyz) [inline src/pyggdrasil/tree_inference/_logprob.py:89]: This is very cool – I think it'll be usable without any modifications in the BlackJAX+SMC version of SCITE :slightly_smiling_face:

File: tests/tree_inference/test_logprob.py (hunk 5 above)
Comment (pawel-czyz) [reply tests/tree_inference/test_logprob.py:124]: Great! :slightly_smiling_face:

**Review by pawel-czyz — APPROVED** [review]
Hi Gordon,

Wonderful work! This implementation looks really good to me and I think we'll be able to use it in SMC version of SCITE powered by BlackJAX without any modifications. Look at it this way – we have prior (uniform over all trees), now you've implemented the likelihood function, and then we'll need the JAX-compatible tree moves to be able to have the very first SMC-powered SCITE version!

Great effort, please merge :slightly_smiling_face:
